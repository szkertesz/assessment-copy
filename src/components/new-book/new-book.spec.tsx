import { render, screen, waitFor } from '@testing-library/react'
import NewBook from './new-book'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { MemoryRouter } from 'react-router-dom'
import userEvent from '@testing-library/user-event'
import thunk from 'redux-thunk'

const mockStore = configureStore([thunk])
let store: any

beforeEach(() => {
    store = mockStore({
        books: {
            books: [],
            status: 'idle',
            error: null,
            filterOptions: {
                status: undefined,
            },
        },
    })
})
describe('NewBook', () => {
    it('should render component', () => {
        const { container } = render(
            <Provider store={store}>
                <MemoryRouter>
                    <NewBook />
                </MemoryRouter>
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    it('should render component', async () => {
        const user = userEvent.setup()
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <NewBook />
                </MemoryRouter>
            </Provider>
        )
        expect(screen.getByLabelText(/Title*/i)).toBeInTheDocument()
        expect(screen.getByLabelText(/Author*/i)).toBeInTheDocument()
        expect(screen.getByLabelText(/Description/i)).toBeInTheDocument()

        await user.type(screen.getByLabelText(/Title*/), 'Test Title')
        await user.type(
            screen.getByRole('textbox', { name: 'Author*' }),
            'Test Author'
        )
        const textarea = screen.getByLabelText(
            /Description/i
        ) as HTMLTextAreaElement
        // textarea.focus()
        await userEvent.type(textarea, 'Test')
        // expect(textarea.value).toBe('Test')
        expect(textarea).toHaveValue('Test')

        const submitButton = await screen.findByRole('button', {
            name: 'Submit',
        })
        await user.click(submitButton)
        const actions = await store.getActions()
        console.log('actions: ', actions)
        expect(actions).toHaveLength(2)
        // expect(await screen.getByText(/Add new book/i)).toBeInTheDocument()
        // const warning = await screen.getByText('Author is required')
        // expect(warning).toBeInTheDocument()
        // await waitFor(() => {
        //     expect(screen.getByText('Author is required')).toBeInTheDocument()
        // })
    })
})
