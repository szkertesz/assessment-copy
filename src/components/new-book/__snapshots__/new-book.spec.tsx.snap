// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewBook should render component 1`] = `
<div>
  <h1>
    Add new book
  </h1>
  <form
    action=""
    class="book-form"
  >
    <div
      class="book-form__group"
    >
      <label
        for="bookTitle"
      >
        Title*
      </label>
      <input
        class="
                            book-form__input
                            "
        id="bookTitle"
        name="bookTitle"
        required=""
        type="text"
      />
    </div>
    <div
      class="book-form__group"
    >
      <label
        for="bookAuthor"
      >
        Author*
      </label>
      <input
        class="
                            book-form__input
                            "
        id="bookAuthor"
        name="bookAuthor"
        required=""
        type="text"
      />
    </div>
    <div
      class="book-form__group"
    >
      <label
        for="bookDescription"
      >
        Description
      </label>
      <textarea
        id="bookDescription"
        name="bookDescription"
        rows="3"
      />
    </div>
    <div
      class="book-form__actions"
    >
      <button
        id="add-book-submit"
        type="submit"
      >
        Submit
      </button>
      <button
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
  <div />
</div>
`;
